src/ioctls/vcpu.rs

```diff
@@ -1045,6 +1045,15 @@ impl VcpuFd {
         Ok(())
     }
 
+    /// Enables guest debugging and sets specific debug registers for x86
+    pub fn set_guest_debug(&self, debug_struct: &kvm_guest_debug) -> Result<()> {
+        let ret = unsafe { ioctl_with_ref(self, KVM_SET_GUEST_DEBUG(), debug_struct) };
```

Comment:
I think we should check the contents for `debug_struct`. The struct can contain different things for different architectures and there's no guarantee that they make sense at this point - for instance the caller could set an `aarch64` specific flag and expect it to work the same on `x86_64`.

Reply:
What exactly do you expect this check to contain?  Also, what flags are you referring to?
If someone calls the `set_guest_debug` from `kvm-iocrtls` on aarch64 and then try to use a field from  `kvm_guest_debug_arch` that is x86_64. Would not they fail the compilation step? What are you meaning by " set an aarch64 specific flag and expect it to work the same on x86_64"?

Reply:
@aghecenco  Are you maybe referring, for example, to the `control` bits supported individually by specific architectures?

Reply:
I hope that KVM will return an error if you're trying to enable debug registers on x86_64 that are only available on aarch64. It might be worth it to have it as a negative test, and check that KVM actually returns an error. Would that make sense?

Reply:
My understanding is that the `kvm_guest_debug_arch`struct is interpreted differently by each architecture (definitions for it differ among architectures) so there seems to be no problem in setting registers - the compiler will complain when using the wrong structure definition. However, regarding the `control` bitfield, when setting KVM_GUESTDBG_EXIT_PENDING (supported on s390 only) for example, and running on x86, no problem would be signaled; it would just have no effect - this is something I have tried.

Reply:
@dianpopa @baciumar yup, I was referring to the `control` field setting flags that mean different things on different architectures. 
* on `aarch64`, the only [valid flags](https://elixir.bootlin.com/linux/v4.20/source/arch/arm64/kvm/guest.c#L459) ([other link](https://elixir.bootlin.com/linux/v4.20/source/arch/arm64/include/uapi/asm/kvm.h#L148))are 0x20000 and 0x10000.
* on `x86_64` the [flags](https://elixir.bootlin.com/linux/v4.20/source/arch/x86/kvm/x86.c#L8421) are much more complex, and many of the `x86_64 flags don't make sense for `aarch64`.

Additionally, there are other flags like @baciumar mentioned, for s390 and so on.
My preference would be for this function to notify the caller when they pass flags that don't make sense, even if KVM ignores them.

Reply:
Ok, now I understand, but are we sure we want to validate kvm-level logic in something that is intended to be a wrapper over low level ioctls? Wouldn t this mean that whenever a new control flag is added for some architecture we would also need to update the logic here?

Reply:
I second what @dianpopa said. We've used kvm-ioclts as a proxy crate that just forwards calls to KVM and checks the result. I think there are cases where we want to do the additional due diligence (like with ever growing input, or with output coming from untrusted actors), but I am not sure it applies here.

Reply:
I agree as well with @dianpopa. This crate is not the right place to validate this very much arch (and in some cases model) specific structure.